lang/lang.go

```diff
@@ -0,0 +1,185 @@
+// or as a map of string keys to any suitable value.
+func LocalizePlural(in string, count int, data ...any) string {
+	return LocalizePluralKey(in, in, count, data...)
+}
+
+// LocalizePluralKey asks the translation engine for the translation with specific ID in plural form.
+// This behaves like the npgettext function, with the `count` parameter determining the plurality looked up.
+// If it cannot be found then the fallback will be used.
+// The string can be templated and the template data can be passed as a struct with exported fields,
+// or as a map of string keys to any suitable value.
+func LocalizePluralKey(key, fallback string, count int, data ...any) string {
+	var d0 any
+	if len(data) > 0 {
+		d0 = data[0]
+	}
+
+	ret, err := localizer.Localize(&i18n.LocalizeConfig{
+		DefaultMessage: &i18n.Message{
+			ID:    key,
+			Other: fallback,
+		},
+		PluralCount:  count,
+		TemplateData: d0,
+	})
+
+	if err != nil {
+		fyne.LogError("Translation failure", err)
+		return fallbackWithData(key, fallback, d0)
+	}
+	return ret
+}
+
+// AddTranslations allows an app to load a bundle of translations.
+// The language that this relates to will be inferred from the resource name, for example "fr.json".
+// The data should be in json format.
+func AddTranslations(r fyne.Resource) error {
+	return addLanguage(r.Content(), r.Name())
+}
+
+// AddTranslationsForLocale allows an app to load a bundle of translations for a specified locale.
+// The data should be in json format.
+func AddTranslationsForLocale(data []byte, l fyne.Locale) error {
+	return addLanguage(data, l.String()+".json")
+}
+
+func addLanguage(data []byte, name string) error {
+	_, err := bundle.ParseMessageFileBytes(data, name)
+	return err
+}
+
+func init() {
+	bundle = i18n.NewBundle(language.English)
+	bundle.RegisterUnmarshalFunc("json", json.Unmarshal)
+
+	loadTranslationsFromFS(translations, "translations")
+
+	// Find the closest translation from the user's locale list and set it up
+	all, err := locale.GetLocales()
+	if err != nil {
+		fyne.LogError("Failed to load user locales", err)
+		all = []string{"en"}
+	}
+	str := closestSupportedLocale(all).LanguageString()
+	setupLang(str)
+	localizer = i18n.NewLocalizer(bundle, str)
+}
+
+func fallbackWithData(key, fallback string, data any) string {
+	t, err := template.New(key).Parse(fallback)
+	if err != nil {
+		log.Println("Could not parse fallback template")
+		return fallback
+	}
+	str := &strings.Builder{}
+	_ = t.Execute(str, data)
+	return str.String()
+}
+
+func loadTranslationsFromFS(fs embed.FS, dir string) {
+	files, err := fs.ReadDir(dir)
+	if err != nil {
+		fyne.LogError("failed to read bundled translations", err)
+		return
+	}
+
+	for _, f := range files {
+		data, err := fs.ReadFile(dir + "/" + f.Name())
+		if err != nil {
+			fyne.LogError("failed to read bundled translation", err)
+			continue
+		}
+		bundle.MustParseMessageFileBytes(data, f.Name())
```

Comment:
This will panic if it can't parse, right? Do we want that?

Reply:
Fixed anyhow, this was getting exposed to developers through the suggested FS public API